# Pandas Intro - Features Analysis


<a id="BasicOSManagement"></a>
## 1. Basic OS Management in Notebook
### Typically ALL the imported packages should be together in a single cell at the top of your notebook

## What is the current working directory (folder) of this notebook and what’s actually in it (a list of files and folders)?
### This will show where you are currently working within your computer's folders. Make sure you know where this folder is. look at the way the path is written in the output that is the form that you need to use. It is different for Windows and Macs

In [7]:
os.listdir()

['.ipynb_checkpoints',
 'Assignment_HealthInsuranceCost_EDA_Reg_Soln.pdf',
 'CO2_DataViz.ipynb',
 'Data',
 'F_MNIST_Data.png',
 'HealthInsuranceCost_EDA_Reg_Soln.ipynb',
 'HW1_PandasIntro_soln.ipynb',
 'HW1_PandasIntro_soln_PDF.pdf',
 'HW1_PandasIntro_std.ipynb',
 'InClassNotes_9_9.ipynb',
 'NBA_Salary_Regressoin_SOLN_v2.pdf',
 'NeuralNetworks_MLP_FashionExp_1.ipynb',
 'Overview Pandas and Partition v1.ipynb',
 'Overview Pandas and Partition v2.ipynb',
 'PandasIntro_Inprogress_Notebook_8_27.ipynb',
 'plot_mnist_filters.ipynb',
 'ProfBaileyNotesPandasIntro.ipynb',
 'Python_regression_results.ipynb',
 'RegressionPractice',
 'Submit.csv',
 'TitanicLogitClassifierAssignment_Soln_PDF.pdf',
 'TitanicLogReg_soln-Copy1.ipynb',
 'TitanicLogReg_soln.ipynb',
 'TitanicRulePredAssignment_soln.ipynb',
 'TitanicRulePredAssignment_Soln_PDF.pdf',
 'Untitled.ipynb',
 'VehiclesEDA_Assignment_Soln.ipynb',
 'VehiclesEDA_Assignment_Soln_PDF.pdf']

## I want to move to a new folder (change directory)! Provide a string with the path.
(tip: If you copy and paste from office or the web, the quote characters may be copied incorrectly (“ vs. ") and that wlll throw an error. Just retype the quotes to fix.)

In [ ]:
os.chdir("C:\\Users\\mdb025\\Desktop")
#can view what's in the folder if wanted
#os.listdir()

## You could also create a path object, check it exists, and move there
### Need to import the function Path from the pathlib package (if you don't want to import an entire package and ALL it's functions, you can just import the ones you need using `from [package name] import [function(s) names]`.

In [16]:
from pathlib import Path
p = Path('C:\\Users\\mdb025\\Documents\\Courses\\Analytics Course Materials\\Machine Learning')
p.exists()

True

In [17]:
#move working directory to this path
os.chdir(p)

<a id="ReadData"></a>
## 1. Read in Data
### Now move to to your original directory (folder) where your Data folder exists and holds the data file `movie.csv`

In [19]:
os.chdir('C:\\Users\\mdb025\\Documents\\Courses\\Analytics Course Materials\\Machine Learning\\ANOP 330\\ANOP330_F2021\\ANOP330F21_Notebooks')

### If the file is in the working directory, then you can just open it by name, otherwise, you may need to add part of the (or the full) path to the file. It's not a bad idea to check the file exists before you try to open it.

In [20]:
os.path.isfile("Data\\movie.csv")

True

### We are going to import using common aliases for our bread and butter packages `pandas` and `numpy`

In [21]:
import pandas as pd
import numpy as np

### Read the data inta a pandas DataFrame (from a downloaded file or even a file from the web). Common convention is to add `df_` as a prefix or `_df` as a suffix to the name for clarity. You'll see both conventions when you look at sample code. Different data file types require different read functions (or options). `read_csv`, `read_excel`, and `read_json` or the most common. There are alternatives for database files also. We'll mostly stick with comma-separated values (CSV) files.
### Whenever we see a new function, do not assume that all you see is all there is, there are MANY optional inputs that you should be exploring! Click and review!
[pandas.read_csv](https://pandas.pydata.org/pandas-docs/dev/reference/api/pandas.read_csv.html)

[pandas.read_excel](https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html)

[pandas.read_json](https://pandas.pydata.org/docs/reference/api/pandas.read_json.html)

In [36]:
df_movies = pd.read_csv("Data\\movie.csv")

#could even read the file from a website if you can find a direct link. 
#This also saves you from having to download and store on your computer (good)
#but you need to have internet access each time your run (bad)
#df_movies = pd.read_csv("https://raw.githubusercontent.com/MattDBailey/ANOP330/main/Data/movie.csv")

### What's in this data?

In [37]:
df_movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
4911  Color        Scott Smith                     1.0      87.0   
4912  Color                NaN                    43.0      43.0   
4913  Color   Benjamin Roberds                    13.0      76.0   
4914  Color        Daniel Hsia                    14.0     100.0   
4915  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
4911                   637.0          NaN                     Comedy|Drama   
4912                   841.0          NaN     Crime|Drama|Mystery|Thriller   
4913                     0.0          NaN            Drama|Horror|Thriller   
4914                   946.0      10443.0             Comedy|Drama|Romance   
4915                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
4911  ...                  6.0  English   Canada             NaN          NaN   
4912  ...                359.0  English      USA           TV-14          NaN   
4913  ...                  3.0  English      USA             NaN       1400.0   
4914  ...                  9.0  English      USA           PG-13          NaN   
4915  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

### Can we get a cleaner view of this data? Look at the top few rows `.head()` or the last few rows `.tail()` or a random sample `.sample()`

In [27]:
df_movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [28]:
df_movies.tail()

color     director_name  num_critic_for_reviews  duration  \
4911  Color       Scott Smith                     1.0      87.0   
4912  Color               NaN                    43.0      43.0   
4913  Color  Benjamin Roberds                    13.0      76.0   
4914  Color       Daniel Hsia                    14.0     100.0   
4915  Color          Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes    gross                        genres  ...  \
4911                   637.0      NaN                  Comedy|Drama  ...   
4912                   841.0      NaN  Crime|Drama|Mystery|Thriller  ...   
4913                     0.0      NaN         Drama|Horror|Thriller  ...   
4914                   946.0  10443.0          Comedy|Drama|Romance  ...   
4915                    86.0  85222.0                   Documentary  ...   

     num_user_for_reviews language  country  content_rating  budget  \
4911                  6.0  English   Canada             NaN     NaN   
4912                359.0  English      USA           TV-14     NaN   
4913                  3.0  English      USA             NaN  1400.0   
4914                  9.0  English      USA           PG-13     NaN   
4915                 84.0  English      USA              PG  1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
4911      2013.0                  470.0        7.7           NaN   
4912         NaN                  593.0        7.5         16.00   
4913      2013.0                    0.0        6.3           NaN   
4914      2012.0                  719.0        6.3          2.35   
4915      2004.0                   23.0        6.6          1.85   

     movie_facebook_likes  
4911                   84  
4912                32000  
4913                   16  
4914                  660  
4915                  456  

[5 rows x 28 columns]

In [31]:
df_movies.sample(5)

color       director_name  num_critic_for_reviews  duration  \
1139  Color     Louis Leterrier                   177.0     103.0   
4430  Color       Mona Fastvold                    16.0      91.0   
1366  Color     Nick Cassavetes                   129.0     116.0   
580   Color        Paul Tibbitt                   147.0      92.0   
4542  Color  Mariette Monpierre                     2.0      78.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
1139                    255.0                  5000.0             Jet Li   
4430                      2.0                     9.0       Brady Corbet   
1366                    415.0                   644.0      Laura Harring   
580                       5.0                   726.0         Billy West   
4542                      0.0                     0.0  Christophe Cherki   

      actor_1_facebook_likes        gross  \
1139                 11000.0   24520892.0   
4430                   321.0          NaN   
1366                 18000.0   71026631.0   
580                    870.0  162495848.0   
4542                     0.0          NaN   

                                         genres  ... num_user_for_reviews  \
1139                Action|Crime|Drama|Thriller  ...                303.0   
4430                                      Drama  ...                  2.0   
1366                       Crime|Drama|Thriller  ...                394.0   
580   Adventure|Animation|Comedy|Family|Fantasy  ...                133.0   
4542                                      Drama  ...                  2.0   

     language  country  content_rating      budget  title_year  \
1139  English   France               R  45000000.0      2005.0   
4430  English   Norway       Not Rated   1100000.0      2014.0   
1366  English      USA           PG-13  36000000.0      2002.0   
580   English      USA              PG  74000000.0      2015.0   
4542   French   France             NaN    900000.0      2011.0   

     actor_2_facebook_likes imdb_score  aspect_ratio movie_facebook_likes  
1139                 5000.0        7.0          2.35                    0  
4430                  287.0        4.9          2.35                  234  
1366                  669.0        7.0          1.85                    0  
580                   861.0        6.0          1.85                16000  
4542                    0.0        6.7         16.00                   24  

[5 rows x 28 columns]

### What are some ATTRIBUTES of the DataFrame?

In [38]:
df_movies.shape

(4916, 28)

In [51]:
#show all the columns and store them
cols = df_movies.columns
cols

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [54]:
#show the index (row names)
ind = df_movies.index
ind

RangeIndex(start=0, stop=4916, step=1)

### See slides for the data dictionary. You should ALWAYS provide your reader a data dictionary (included created fields). The correctness of your analysis is COMPLETELY dependent on the what each data field (and row) represents! It's the data is the product for building, so you better understand your product.

## What are the types of data?
### pandas tries to guess what the data types are, so check that it guessed correctly (dates as datetime, integers as integers, etc.). Watch out for numerical fields that are actually categorical (or objects). Such as 0=Monday, 1=Tuesday, etc.. See slides for types.

In [58]:
df_movies.dtypes

color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
movie_title                   object
num_voted_users                int64
cast_total_facebook_likes      int64
actor_3_name                  object
facenumber_in_poster         float64
plot_keywords                 object
movie_imdb_link               object
num_user_for_reviews         float64
language                      object
country                       object
content_rating                object
budget                       float64
title_year                   float64
actor_2_facebook_likes       float64
imdb_score                   float64
aspect_ratio                 float64
m

In [69]:
df_movies['content_rating'].value_counts()

R            2067
PG-13        1411
PG            686
Not Rated     115
G             112
Unrated        59
Approved       54
TV-14          30
TV-MA          18
TV-PG          13
X              12
TV-G           10
Passed          9
NC-17           7
GP              6
M               5
TV-Y7           1
TV-Y            1
Name: content_rating, dtype: int64

In [71]:
df_movies['content_rating'].astype('category').cat.categories

Index(['Approved', 'G', 'GP', 'M', 'NC-17', 'Not Rated', 'PG', 'PG-13',
       'Passed', 'R', 'TV-14', 'TV-G', 'TV-MA', 'TV-PG', 'TV-Y', 'TV-Y7',
       'Unrated', 'X'],
      dtype='object')

## Let's get some information about the fields (columns)

In [72]:
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      4897 non-null   object 
 1   director_name              4814 non-null   object 
 2   num_critic_for_reviews     4867 non-null   float64
 3   duration                   4901 non-null   float64
 4   director_facebook_likes    4814 non-null   float64
 5   actor_3_facebook_likes     4893 non-null   float64
 6   actor_2_name               4903 non-null   object 
 7   actor_1_facebook_likes     4909 non-null   float64
 8   gross                      4054 non-null   float64
 9   genres                     4916 non-null   object 
 10  actor_1_name               4909 non-null   object 
 11  movie_title                4916 non-null   object 
 12  num_voted_users            4916 non-null   int64  
 13  cast_total_facebook_likes  4916 non-null   int64

## Here is a function that I found to decrease the memory of your data set by picking a smaller data type (e.g., 8 bit float from 64 bit float)
## It isn't much of an issue for our in-class data, but handy for larger data sets.

In [73]:
# Function to decrease data usage in notebook (from Baek Kyun Shin)
def downcast(df, verbose=True):
    start_mem = df.memory_usage().sum() / 1024**2
    for col in df.columns:
        dtype_name = df[col].dtype.name
        if dtype_name == 'object':
            pass
        elif dtype_name == 'bool':
            df[col] = df[col].astype('int8')
        elif dtype_name.startswith('int') or (df[col].round() == df[col]).all():
            df[col] = pd.to_numeric(df[col], downcast='integer')
        else:
            df[col] = pd.to_numeric(df[col], downcast='float')
    end_mem = df.memory_usage().sum() / 1024**2
    if verbose:
        print('{:.1f}% Compressed'.format(100 * (start_mem - end_mem) / start_mem))
    
    return df

In [75]:
df_movies = downcast(df_movies)

28.6% Compressed


In [76]:
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      4897 non-null   object 
 1   director_name              4814 non-null   object 
 2   num_critic_for_reviews     4867 non-null   float32
 3   duration                   4901 non-null   float32
 4   director_facebook_likes    4814 non-null   float32
 5   actor_3_facebook_likes     4893 non-null   float32
 6   actor_2_name               4903 non-null   object 
 7   actor_1_facebook_likes     4909 non-null   float32
 8   gross                      4054 non-null   float32
 9   genres                     4916 non-null   object 
 10  actor_1_name               4909 non-null   object 
 11  movie_title                4916 non-null   object 
 12  num_voted_users            4916 non-null   int32  
 13  cast_total_facebook_likes  4916 non-null   int32

## Reduced from 1.1 MB (1100 KB) to 0.768 MB (768 KB) a 28.6% reduction in memory! Mostly from moving from float64 and int64 to float32 and int32.

## Next we'll look at working with a single column (series) from the dataframe.